### PR TITLE
Exception-free, branch-free PageCursor bounds-checks.

### PR DESCRIPTION
**Pull request by chrisvest**
Performing an out-of-bounds memory access to a page via a page cursor now no longer throws an exception.
The reason is that we might have an inconsistent read that fools decoding logic computing wrong offsets for data, or to try and read more data than can fit on a page.
Previously, such out-of-bounds accesses would cause the PageCursor to throw an exception, but this is inconsiderate - it is not the fault of the decoding logic that it was fed wrong data and tricked into an out-of-bounds access.
So instead we now raise a boolean flag the cursor has experienced and out-of-bounds access, and this flag can then be checked after the decoder has checked that its read was consistent.
Then it is up to the decoder to make a decision about how to handle the fact that it did an out-of-bounds access.
At present, the record formats (where the decoding logic is) don't do anything, but this is just because this change is not yet complete.

@magnusvejlstrup I think this should go into 3.0, because we cannot otherwise forward merge the more devious adversarial page cache rule - and neatly fix the problems it uncovers - without this. The change is somewhat risky, though. I think the build most likely to catch any errors introduced by this, is our page cache stress robustness build.

### CODE REVIEW

**Comment by chrisvest** [comment]
Marked as "NOT READY FOR MERGE" because I still haven't made the record formats do something reasonable when they do out-of-bounds access.

File: community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java [hunk 1]
```diff
@@ -325,33 +331,28 @@ protected long assertPagedFileStillMappedAndGetIdOfLastPage()
 
     // --- IO methods:
 
-    private void checkBounds( int position )
+    /**
+     * Compute a pointer that guarantees (assuming {@code size} is less than or equal to {@link #pageSize}) that the
+     * page access will be within the bounds of the page.
+     * This might mean that the pointer won't point to where one might naively expect, but will instead be
+     * truncated to point within the page. In this case, an overflow has happened anf the {@link #outOfBounds}
```
Comment (MishaDemianenko) [inline community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java]: anf -> and

**Comment by MishaDemianenko** [comment]
@chrisvest that's it for now, waiting for remaining part

**Comment by chrisvest** [comment]
@MishaDemianenko HAR build is also green, so this looks good.

File: community/io/src/test/java/org/neo4j/io/pagecache/impl/CompositePageCursorTest.java [hunk 2]
```diff
@@ -0,0 +1,1149 @@
...
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.neo4j.test.ByteArrayMatcher.byteArray;
+
+//@RunWith( Theories.class )
```
Comment (MishaDemianenko) [inline community/io/src/test/java/org/neo4j/io/pagecache/impl/CompositePageCursorTest.java:38]: please remove it in some of the nearest future commits

Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/io/pagecache/impl/CompositePageCursorTest.java:38]: Oh, yeah, that one slipped through the cracks

**Pull request merged**